File: CSharpCodeBase/aiactions/ai_followparent.lua
local Class = require 'utils.hump.class'
local AINode = require 'utils.ailib.ainode'

-- Move to enemy
local AI_FollowParent = Class({__includes=AINode, init = function (self, entity)
  AINode.init(self, entity)
  self.radiusMax = self.entity.config:Get("parentSafeRadiusMax")
  self.radiusMin = self.entity.config:Get("parentSafeRadiusMin")
end})

function AI_FollowParent:Visit()
  local parent = self.entity.relationship:GetInstance("parent")
  if self.status == NODE_READY then
    if self.entity:GetSimpleDistance(parent) < self.radiusMax then
      -- nothing to do, we are in safe-zone
      self.status = NODE_SUCCESS
    else
      -- we need move to parent
      self.status = NODE_RUNNING
    end
  elseif self.status == NODE_RUNNING then
    if parent == nil then 
      self.status = NODE_FAILURE
    elseif self.entity:GetSimpleDistance(parent) > self.radiusMin then
      self.entity.mover:ResetSpeed()
      self.entity.mover:SetInputToVec(parent:GetPosition())
    else
      self.status = NODE_SUCCESS
    end
  end
end

return AI_FollowParent

File: CSharpCodeBase/aiactions/ai_followparent.cs
using System;
 
 using System.Collections.Generic;
 
 using System.Linq;
 
 using System.Text;
 
 using UnityEngine;
 
 using System.Collections;
 
 using MainGame.core;
 
 using UnityEngine.EventSystems;
 
 namespace MainGame{
 public class ai_followparent :ainode{
 // Move to enemy
 public void init (self, entity);
   AINode.init(self, entity);
   this.radiusMax = this.entity.config:Get("parentSafeRadiusMax");
   this.radiusMin = this.entity.config:Get("parentSafeRadiusMin");
 }})
 public void Visit(){
   var parent = this.entity.relationship:GetInstance("parent");
   if(this.status == NODE_READY  ){ 
     if(this.entity:GetSimpleDistance(parent) < this.radiusMax  ){ 
       // nothing to ){, we are in safe-zone
       this.status = NODE_SUCCESS;
     }else{
       // we need move to parent
       this.status = NODE_RUNNING;
     }
   }else{if this.status == NODE_RUNNING  ){ 
     if(parent == null  ){  
       this.status = NODE_FAILURE;
     }else{if this.entity:GetSimpleDistance(parent) > this.radiusMin  ){ 
       this.entity.mover:ResetSpeed();
       this.entity.mover:SetInputToVec(parent:GetPosition());
     }else{
       this.status = NODE_SUCCESS;
     }
   }
 }
 }}